style: apply /style-guide pass to models/tables (#2725)
wandb/docs: models/tables/log_tables.mdx

---
title: Log tables
description: "Create and log W&B Tables with different logging modes including immutable, mutable, and incremental using the Python SDK."
keywords: ["wandb.Table", "log_mode", "add_column", "add_data", "batch logging"]
---

This page shows you how to create and log W&B Tables with the Python SDK so you can visualize and analyze tabular data, including predictions, evaluation results, and batched training output, alongside your ML experiments.

A W&B Table is a two-dimensional grid of data where each column has a single type of data. Each row represents one or more data points logged to a W&B [run](/models/runs/). W&B Tables support primitive and numeric types, as well as nested lists, dictionaries, and rich media types.

A W&B Table is a specialized [data type](/models/ref/python/data-types/) in W&B, logged as an [artifact](/models/artifacts/) object.

> You [create and log table objects](#create-and-log-a-new-table) using the W&B Python SDK. When you create a table object, you specify the columns and data for the table and a [mode](#table-logging-modes). The mode determines how the table is logged and updated during your ML experiments, which affects performance, what you can change after logging, and how the table appears in the W&B App.

Author: Fixed in 7910ed9. Updated the two intro links to `#create-and-log-a-table` and `#logging-modes` so they match the headings `## Create and log a table` and `## Logging modes`. (Mintlify strips backticks and lowercases/hyphenates, so `## Logging modes` → `#logging-modes`, which the line-25 link already used.)

---

> This page shows you how to create and log W&B Tables with the Python SDK so you can visualize and analyze tabular data, including predictions, evaluation results, and batched training output, alongside your ML experiments.

Reviewer: ```suggestion
This page shows you how to create and log W&B Tables with the Python SDK so you can visualize and analyze tabular data alongside your ML experiments.
```

> A Table is a two-dimensional grid of data where each column has a single type of data. Each row represents one or more data points logged to a W&B [run](/models/runs/). Tables support primitive and numeric types, as well as nested lists, dictionaries, and rich media types.

Reviewer: Why does it keep removing "W&B"?
```suggestion
A *table* is a two-dimensional grid of data where each column has a single type of data. Each row represents one or more data points logged to a [run](/models/runs/). Tables support primitive and numeric types, as well as nested lists, dictionaries, and rich media types.
```

> A Table is a specialized W&B [data type](/models/ref/python/data-types/). When you log a table, W&B stores it as an [artifact](/models/artifacts/).

Reviewer: ```suggestion
A W&B Table is a specialized W&B [data type](/models/ref/python/data-types/). When you log a table, W&B stores it as an [artifact](/models/artifacts/).
```

You [create and log table objects](#create-and-log-a-new-table) using the W&B Python SDK. When you create a table object, you specify the columns and data for the table and a [mode](#table-logging-modes). The mode determines how the table is logged and updated during your ML experiments.
A Table is a specialized W&B [data type](/models/ref/python/data-types/). When you log a table, W&B stores it as an [artifact](/models/artifacts/).

> <Note>

Reviewer: Do not remove this.

Author: Matt specifically suggests removal since any W&B Server version below this threshhold is EOL, which I agree with.

> Follow these steps to log a table to a run. The resulting table is stored as an artifact in W&B and rendered in the run's workspace.

Reviewer: ```suggestion
Follow these steps to log a table to a run. The resulting table is stored as an artifact in W&B and rendered in your project's workspace.
```

1. Initialize a new run with `wandb.init()`.

> 2. Create a Table object with the [`wandb.Table`](/models/ref/python/data-types/table) class. Specify the columns and data for the table for the `columns` and `data` parameters, respectively. Set the optional `log_mode` parameter to one of the three modes: `IMMUTABLE` (the default), `MUTABLE`, or `INCREMENTAL`, because the mode controls how the table behaves when logged. See [Logging modes](#logging-modes) for more information.

Reviewer: This reason why "Table" is correct here is because it refers to an instance of a Table (Python) Class.

I think the bot is confusing product names with how to refer to Python objects.
```suggestion
2. Create a table with the [`wandb.Table`](/models/ref/python/data-types/table) Class. Use the `columns` and `data` parameters to define the table columns and rows. To control how the table behaves when logged, set the optional `log_mode` parameter to: `IMMUTABLE` (the default), `MUTABLE`, or `INCREMENTAL`. See [Table Logging Modes](#logging-modes) in the next section for more information.
```

---

This page shows you how to create and log W&B Tables with the Python SDK so you can visualize and analyze tabular data, including predictions, evaluation results, and batched training output, alongside your ML experiments.

> A W&B Table is a two-dimensional grid of data where each column has a single type of data. Each row represents one or more data points logged to a W&B [run](/models/runs/). W&B Tables support primitive and numeric types, as well as nested lists, dictionaries, and rich media types.

Reviewer 2: ```suggestion
A Table is a two-dimensional grid of data where each column has a single type of data. Each row represents one or more data points logged to a W&B [run](/models/runs/). Tables support primitive and numeric types, as well as nested lists, dictionaries, and rich media types.
```
Scrub for extraneous "W&B" instances beyond the first mention

Author: Applied; also scrubbed "W&B Table(s)" to "Table(s)" beyond the first mention on this page.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

A W&B Table is a specialized [data type](/models/ref/python/data-types/) in W&B, logged as an [artifact](/models/artifacts/) object.

> You [create and log table objects](#create-and-log-a-table) using the W&B Python SDK. When you create a table object, you specify the columns and data for the table and a [mode](#logging-modes). The mode determines how the table is logged and updated during your ML experiments, which affects performance, what you can change after logging, and how the table appears in the W&B App.

Reviewer 2: ```suggestion
You [create and log table objects](#create-and-log-a-table) using the W&B Python SDK. When you create a table object, you specify the columns and data for the table and a [mode](#logging-modes). The mode determines how the table is logged and updated during your ML experiments, which affects performance, what you can change after logging, and how the table appears in W&B.
```

Author: Changed "the W&B App" to "W&B" here and in the other descriptive mentions on the page.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

<Note>

> `INCREMENTAL` mode is supported on W&B Server v0.70.0 and above.

Reviewer 2: I think this note can be dropped because earlier versions are EOL.

Author: Dropped the note.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

> </Note>

Author: ```suggestion
```

This page shows you how to create and log W&B Tables with the Python SDK so you can visualize and analyze tabular data, including predictions, evaluation results, and batched training output, alongside your ML experiments.

A Table is a two-dimensional grid of data where each column has a single type of data. Each row represents one or more data points logged to a W&B [run](/models/runs/). Tables support primitive and numeric types, as well as nested lists, dictionaries, and rich media types.

A Table is a specialized [data type](/models/ref/python/data-types/) in W&B, logged as an [artifact](/models/artifacts/) object.

> You [create and log table objects](#create-and-log-a-table) using the W&B Python SDK. When you create a table object, you specify the columns and data for the table and a [mode](#logging-modes). The mode determines how the table is logged and updated during your ML experiments, which affects performance, what you can change after logging, and how the table appears in W&B.

Reviewer 2: Maybe a nit, but now I'm wondering what's the difference between logging a Table artifact (line 11) and a table object (line 13). If any?

Author: Good catch — reworded to remove the ambiguity. There is really one object: a table is a W&B data type, and when you log a table W&B stores it *as* an artifact (not a separate "artifact object"). Also switched line 13 from "table objects" to "tables" so the terminology is consistent.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

## Logging modes

> The logging mode you choose affects performance, what you can change after logging, and how the table appears in W&B. Pick the mode that matches your workflow, for example, a single end-of-run snapshot, a table you progressively enrich with new columns, or a long-running training table updated in batches.

Reviewer: @johndmulhausen This is a good example of the biggest pain point I am finding reviewing these PRs. TL;DR too wordy and it recycles existing content.

I asked ChatGPT this (I told it to use the Google Developer style guide):

Prompt
"I want you to look at what a human wrote vs an LLM. Tell me which is better and also propose your own fix:"

Answer
"The human version is better as a docs passage because it is direct, scoped, and not repetitive. The bot version adds useful context, but it duplicates the human paragraph and introduces more words than needed."

> The [`wandb.Table`](/models/ref/python/data-types/table) `log_mode` parameter determines how a table is logged and updated during your ML experiments. The `log_mode` parameter accepts one of three arguments: `IMMUTABLE`, `MUTABLE`, and `INCREMENTAL`. Each mode has different implications for how a table is logged, how it can be modified, and how it is rendered in W&B.

Reviewer: ```suggestion
The [`wandb.Table`](/models/ref/python/data-types/table) `log_mode` parameter controls how a table is logged, updated, and displayed during in the W&B App. Choose one of the following values based on your workflow: `IMMUTABLE` for a single end-of-run snapshot, `MUTABLE` for a table that you update after logging, or `INCREMENTAL` for long-running jobs that log table data in batches.
```

## Logging modes

> The logging mode you choose affects performance, what you can change after logging, and how the table appears in the W&B App. Pick the mode that matches your workflow, for example, a single end-of-run snapshot, a table you progressively enrich with new columns, or a long-running training table updated in batches.

Reviewer 2: ```suggestion
The logging mode you choose affects performance, what you can change after logging, and how the table appears in W&B. Pick the mode that matches your workflow, for example, a single end-of-run snapshot, a table you progressively enrich with new columns, or a long-running training table updated in batches.
```

Author: Changed "the W&B App" to "W&B".

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

The [`wandb.Table`](/models/ref/python/data-types/table) `log_mode` parameter determines how a table is logged and updated during your ML experiments. The `log_mode` parameter accepts one of three arguments: `IMMUTABLE`, `MUTABLE`, and `INCREMENTAL`. Each mode has different implications for how a table is logged, how it can be modified, and how it is rendered in the W&B App.

The following describes the three logging modes, the high-level differences, and common use cases for each mode:

| Mode  | Definition | Use cases  | Benefits  |
| ----- | ---------- | ---------- | ----------|
| `IMMUTABLE`   | Once a table is logged to W&B, you cannot modify it. |- Storing tabular data generated at the end of a run for further analysis                              | - Minimal overhead when logged at the end of a run<br/>- All rows rendered in UI |

> | `MUTABLE`     | After you log a table to W&B, you can overwrite the existing table with a new one. | - Adding columns or rows to existing tables<br/>- Enriching results with new information                        | - Capture table mutations<br/>- All rows rendered in UI                          |

Reviewer 2: The list here son't be a real list. Consider reformatting as a `<ul>`.

Author: Reformatted the Use cases and Benefits cells as <ul> lists in all three rows.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

    run.log({"eval_table": table})  # Final update with another column
```

If you only want to add new batches of rows (no columns) incrementally like in a training loop, consider using [`INCREMENTAL` mode](#incremental-mode) instead.

### `INCREMENTAL` mode

In `INCREMENTAL` mode, you log batches of rows to a table during the machine learning experiment. This is ideal for monitoring long-running jobs or when working with large tables that would be inefficient to log during the run for updates. Within the UI, the table is updated with new rows as they are logged, so you can view the latest data without waiting for the entire run to finish. You can also step through the increments to view the table at different points in time.

<Note>
Run workspaces in W&B have a limit of 100 increments. If you log more than 100 increments, only the most recent 100 are shown in the run workspace.
</Note>

> The following example creates a table in `INCREMENTAL` mode, logs it, and then adds new rows to it. The table is logged once per training step (`step`). The example uses placeholder functions (`get_training_batch()`, `train_model_on_batch()`, and `predict_on_batch()`); replace these with your own data loading, training, and prediction functions.

Reviewer: I'd avoid using semicolons. Re: That'll make the readability (esp for ESL) more difficult.

    run.log({"eval_table": table})  # Final update with another column
```

If you only want to add new batches of rows (no columns) incrementally like in a training loop, consider using [`INCREMENTAL` mode](#incremental-mode) instead.

### `INCREMENTAL` mode

In `INCREMENTAL` mode, you log batches of rows to a table during the machine learning experiment. This is ideal for monitoring long-running jobs or when working with large tables that would be inefficient to log during the run for updates. Within the UI, the table is updated with new rows as they are logged, so you can view the latest data without waiting for the entire run to finish. You can also step through the increments to view the table at different points in time.

<Note>
Run workspaces in the W&B App have a limit of 100 increments. If you log more than 100 increments, only the most recent 100 are shown in the run workspace.
</Note>

The following example creates a table in `INCREMENTAL` mode, logs it, and then adds new rows to it. The table is logged once per training step (`step`).

> <Note>

Reviewer 2: Consider rewriting to drop one of these notes, which reduce the visibility of the para at line 118.

Author: Inlined the placeholder-functions note into the paragraph, so a single note no longer sandwiches it.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
